tests/test_cache_api.py

```diff
@@ -0,0 +1,175 @@
+import pytest
+
+import pickle
+
+import bionic as bn
+from bionic import interpret
+from bionic.persistence import (
+    path_from_url,
+    is_file_url,
+    is_gcs_url,
+    bucket_and_object_names_from_gs_url,
+)
+from bionic.util import get_gcs_client_without_warnings
+
+
+class CacheTester:
+    """
+    A helper class for testing changes to Bionic's cache.
+
+    Tracks the current and previous states of the Cache API, allow us to express tests
+    in terms of changes between states.
+    """
+
+    def __init__(self, flow, tier=["local", "cloud"]):
+        self.flow = flow
+        self._old_entries = set()
+
+        self._tiers = interpret.str_or_seq_as_list(tier)
+
+    def expect_new_entries(self, *expected_new_entity_names):
+        cur_entries = set(self._get_entries())
+        assert cur_entries.issuperset(self._old_entries)
+        new_entries = set(cur_entries) - self._old_entries
+        self._old_entries = cur_entries
+
+        new_entity_names = {entry.entity for entry in new_entries}
+        assert new_entity_names == set(expected_new_entity_names)
+
+        for entry in new_entries:
+            self._validate_entry(entry)
+
+    def expect_zero_entries(self):
+        assert list(self._get_entries()) == []
+
+    def _get_entries(self):
+        return [
+            entry
+            for entry in self.flow.cache.get_entries()
+            if entry.tier in self._tiers
+        ]
+
+    def _validate_entry(self, entry):
+        artifact_bytes = read_bytes_from_url(entry.artifact_url)
+        value = pickle.loads(artifact_bytes)
+        assert value == self.flow.get(entry.entity)
+
+        if entry.tier == "local":
+            artifact_path_bytes = entry.artifact_path.read_bytes()
+            assert artifact_path_bytes == artifact_bytes
+        else:
+            assert entry.artifact_path is None
+
+        # We won't make too many assumptions about the format of the metadata, but we
+        # can check that it contains the entity name. (Unfortunately it won't
+        # necessarily contain the absolute artifact URL; it may be a relative URL
+        # instead.)
+        metadata_str = read_bytes_from_url(entry.metadata_url).decode("utf-8")
```

Comment:
It does not have to be in this PR, but `CacheEntry` is a good candidate for unit tests. We can test properties easily using unit tests that would be complicated to test here.

Reply:
What properties did you have in mind here? The things like the comparison operators?

(I'm a little skeptical of adding unit tests for this class because its interface isn't completely stable or user-visible: its properties and methods are stable, but its constructor is coupled to Bionic's internals and will probably keep changing. For example, if we do decide to merge `ExternalCacheItem` with `InventoryEntry`, we'll have to modify these unit tests.)